File: src/typescript/src/poly-bus.ts
import { IPolyBus } from './i-poly-bus';
import { ITransport } from './transport/i-transport';
import { IncomingHandler } from './transport/transaction/message/handlers/incoming-handler';
import { OutgoingHandler } from './transport/transaction/message/handlers/outgoing-handler';
import { Messages } from './transport/transaction/message/messages';
import { Transaction } from './transport/transaction/transaction';
import { IncomingTransaction } from './transport/transaction/incoming-transaction';
import { OutgoingTransaction } from './transport/transaction/outgoing-transaction';
import { IncomingMessage } from './transport/transaction/message/incoming-message';
import { PolyBusBuilder } from './poly-bus-builder';

/**
 * Implementation of IPolyBus that provides message handling and transport functionality.
 */
export class PolyBus implements IPolyBus {
  private _transport!: ITransport;
  private readonly _incomingPipeline: IncomingHandler[];
  private readonly _outgoingPipeline: OutgoingHandler[];
  private readonly _messages: Messages;
  private readonly _name: string;
  private readonly _builder: PolyBusBuilder;

  /**
   * Creates a new PolyBus instance.
   * @param builder The builder containing configuration for this bus instance.
   */
  constructor(builder: PolyBusBuilder) {
    this._builder = builder;
    this._incomingPipeline = builder.incomingPipeline;
    this._outgoingPipeline = builder.outgoingPipeline;
    this._messages = builder.messages;
    this._name = builder.name;
  }

  /**
   * The properties associated with this bus instance.
   */
  public get properties(): Map<string, object> {
    return this._builder.properties;
  }

  /**
   * The transport mechanism used by this bus instance.
   */
  public get transport(): ITransport {
    return this._transport;
  }

  public set transport(value: ITransport) {
    this._transport = value;
  }

  /**
   * Collection of handlers for processing incoming messages.
   */
  public get incomingPipeline(): IncomingHandler[] {
    return this._incomingPipeline;
  }

  /**
   * Collection of handlers for processing outgoing messages.
   */
  public get outgoingPipeline(): OutgoingHandler[] {
    return this._outgoingPipeline;
  }

  /**
   * Collection of message types and their associated headers.
   */
  public get messages(): Messages {
    return this._messages;
  }

  /**
   * The name of this bus instance.
   */
  public get name(): string {
    return this._name;
  }

  /**
   * Creates a new incoming transaction.
   * @param message Incoming message to create the transaction from.
   * @returns A promise that resolves to the created incoming transaction.
   */
  public async createIncomingTransaction(message: IncomingMessage): Promise<IncomingTransaction> {
    return this._builder.incomingTransactionFactory(this._builder, this, message);
  }

  /**
   * Creates a new outgoing transaction.
   * @returns A promise that resolves to the created outgoing transaction.
   */
  public async createOutgoingTransaction(): Promise<OutgoingTransaction> {
    return this._builder.outgoingTransactionFactory(this._builder, this);
  }

  /**
   * Sends messages associated with the given transaction to the transport.
   * Applies appropriate handlers based on transaction type before sending.
   * @param transaction The transaction containing messages to send.
   * @returns A promise that resolves when the messages have been sent.
   */
  public async send(transaction: Transaction): Promise<void> {
    let step = () => this.transport.handle(transaction);

    if (transaction instanceof IncomingTransaction) {
      const handlers = transaction.bus.incomingPipeline;
      for (let index = handlers.length - 1; index >= 0; index--) {
        const handler = handlers[index];
        const next = step;
        step = () => handler(transaction, next);
      }
    } else if (transaction instanceof OutgoingTransaction) {
      const handlers = transaction.bus.outgoingPipeline;
      for (let index = handlers.length - 1; index >= 0; index--) {
        const handler = handlers[index];
        const next = step;
        step = () => handler(transaction, next);
      }
    }

    try {
      await step();
    } catch (error) {
      await transaction.abort();
      throw error;
    }
  }

  /**
   * Starts the bus and begins processing messages.
   * @returns A promise that resolves when the bus has started.
   */
  public async start(): Promise<void> {
    return this.transport.start();
  }

  /**
   * Stops the bus and ceases processing messages.
   * @returns A promise that resolves when the bus has stopped.
   */
  public async stop(): Promise<void> {
    return this.transport.stop();
  }
}

File: src/python/src/poly_bus.py
"""PolyBus implementation for the Python version."""

from typing import Dict, Any, List
from .i_poly_bus import IPolyBus
from .transport.i_transport import ITransport
from .transport.transaction.incoming_transaction import IncomingTransaction
from .transport.transaction.outgoing_transaction import OutgoingTransaction
from .transport.transaction.message.incoming_message import IncomingMessage
from .transport.transaction.message.messages import Messages
from .transport.transaction.message.handlers.incoming_handler import IncomingHandler
from .transport.transaction.message.handlers.outgoing_handler import OutgoingHandler
from .transport.transaction.transaction import Transaction


class PolyBus(IPolyBus):
    """Main PolyBus implementation that provides message handling and transport functionality."""
    
    def __init__(self, builder):
        """
        Initialize a PolyBus instance from a builder.
        
        Args:
            builder: The PolyBusBuilder containing the configuration
        """
        self._builder = builder
        self._transport: ITransport = None  # type: ignore
    
    @property
    def properties(self) -> Dict[str, Any]:
        """The properties associated with this bus instance."""
        return self._builder.properties
    
    @property
    def transport(self) -> ITransport:
        """The transport mechanism used by this bus instance."""
        return self._transport
    
    @transport.setter
    def transport(self, value: ITransport) -> None:
        """Set the transport mechanism used by this bus instance."""
        self._transport = value
    
    @property
    def incoming_pipeline(self) -> List[IncomingHandler]:
        """Collection of handlers for processing incoming messages."""
        return self._builder.incoming_pipeline
    
    @property
    def outgoing_pipeline(self) -> List[OutgoingHandler]:
        """Collection of handlers for processing outgoing messages."""
        return self._builder.outgoing_pipeline
    
    @property
    def messages(self) -> Messages:
        """Collection of message types and their associated headers."""
        return self._builder.messages
    
    @property
    def name(self) -> str:
        """The name of this bus instance."""
        return self._builder.name
    
    @property
    def incoming_handlers(self) -> List[IncomingHandler]:
        """Backwards-compatible alias for incoming_pipeline."""
        return self.incoming_pipeline
    
    @property
    def outgoing_handlers(self) -> List[OutgoingHandler]:
        """Backwards-compatible alias for outgoing_pipeline."""
        return self.outgoing_pipeline
    
    async def create_transaction(self, message: IncomingMessage = None):
        """Backwards-compatible method to create transactions.
        
        Args:
            message: Optional incoming message. If provided, creates an incoming transaction.
                    If not provided, creates an outgoing transaction.
                    
        Returns:
            Either an IncomingTransaction or OutgoingTransaction depending on whether message is provided.
        """
        if message is not None:
            return await self.create_incoming_transaction(message)
        else:
            return await self.create_outgoing_transaction()
    
    async def create_incoming_transaction(self, message: IncomingMessage) -> IncomingTransaction:
        """Creates a new incoming transaction based on an incoming message.
        
        Args:
            message: The incoming message to create the transaction from.
            
        Returns:
            A new incoming transaction instance.
        """
        return await self._builder.incoming_transaction_factory(self._builder, self, message)
    
    async def create_outgoing_transaction(self) -> OutgoingTransaction:
        """Creates a new outgoing transaction.
            
        Returns:
            A new outgoing transaction instance.
        """
        return await self._builder.outgoing_transaction_factory(self._builder, self)
    
    async def send(self, transaction: Transaction) -> None:
        """Sends messages associated with the given transaction to the transport.
        
        Args:
            transaction: The transaction containing messages to send.
        """
        async def final_step() -> None:
            """Final step that actually sends to transport."""
            await self.transport.handle(transaction)
        
        step = final_step
        
        # Build handler chain based on transaction type
        if isinstance(transaction, IncomingTransaction):
            # Process incoming handlers in reverse order
            handlers = transaction.bus.incoming_pipeline
            for index in range(len(handlers) - 1, -1, -1):
                handler = handlers[index]
                next_step = step
                
                # Fix closure issue by using default parameters
                def create_step(h: IncomingHandler = handler, next_fn = next_step) -> Any:
                    async def handler_step() -> None:
                        await h(transaction, next_fn)
                    return handler_step
                
                step = create_step()
        
        elif isinstance(transaction, OutgoingTransaction):
            # Process outgoing handlers in reverse order
            handlers = transaction.bus.outgoing_pipeline
            for index in range(len(handlers) - 1, -1, -1):
                handler = handlers[index]
                next_step = step
                
                # Fix closure issue by using default parameters
                def create_step(h: OutgoingHandler = handler, next_fn = next_step) -> Any:
                    async def handler_step() -> None:
                        await h(transaction, next_fn)
                    return handler_step
                
                step = create_step()
        
        try:
            await step()
        except Exception:
            await transaction.abort()
            raise
    
    async def start(self) -> None:
        """Enables the bus to start processing messages."""
        await self.transport.start()
    
    async def stop(self) -> None:
        """Stops the bus from processing messages."""
        await self.transport.stop()
